### ISSUE TITLE
Kurtossis (and maybe Skewness) metafeatures

### ISSUE THREAD

**Issue by PG-TUe** [openml/OpenML#297]
Hi,

I have been working on computing my own meta-features, and when I tried to verify my Kurtosis and Skewness meta-features with those on the website (in particular for the iris dataset: openml.org/d/61) I came up with different results.

For the iris dataset, the mean of kurtosises (MeanKurtosisOfNumeric) has a value of 32.79 which seems extremely high to me. Through using scipy.stats.kurtosis, I get the following kurtosis values: -0.57, 0.24, -1.40, -1.33. The mean of these values is of course -0.77, nowhere close to 32.79. I looked at the code that calculates the kurtosis, but I couldn't find any mistakes from glancing over quickly. I checked with Excel, and while they use a slightly different version of kurtosis, their values were also very similar to mine.

My skewnesses results also differ, again I checked with excel and scipy, but the skewness on OpenML might not be entirely correct either. The values differ less though (my/excel skewness mean is roughly 0.06, while OpenML says 1.6), but it still seems like a big difference to me.

Any input is appreciated.

**Comment by berndbischl** [comment]
the kurtosis can never be < 0?

see here 
https://en.wikipedia.org/wiki/Kurtosis

this is what R computes

```
library(moments)
d = iris[, -5]
k = kurtosis(d)
print(k)
print(mean(k))
```

```
Sepal.Length  Sepal.Width Petal.Length  Petal.Width 
       2.426        3.181        1.604        1.664 

[1] 2.219
```
 
which of course also completely different to the OML numbers you quoted ;-)

**Comment by berndbischl** [comment]
ok, scipy computes the excess kurtosis by fisher, so we have to subtract 3.... 

```
> k - 3
Sepal.Length  Sepal.Width Petal.Length  Petal.Width 
     -0.5736       0.1810      -1.3955      -1.3361 
 ```

**Comment by berndbischl** [comment]
now we (me and @PG-TUe ) still have problem with the 2nd value? why is that different?

**Comment by berndbischl** [comment]
cool, apparently the iris data set from R (which i used) and the one on OML are not the same...

```

> j = c(35, 38)

# from R
> iris[j, ]
   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
35          4.9         3.1          1.5         0.2  setosa
38          4.9         3.6          1.4         0.1  setosa

# from OML / did = 61
> od$data[j, ]
   sepallength sepalwidth petallength petalwidth       class
34         4.9        3.1         1.5        0.1 Iris-setosa
37         4.9        3.1         1.5        0.1 Iris-setosa

```


paging @joaquinvanschoren to this horrible thread

**Comment by PG-TUe** [comment]
Yes indeed, scipy.stats uses Fisher by default, sorry for that omission ([link to scipy docs](http://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.kurtosis.html)).

Different datasets makes sense! It wouldn't be anything in the computation if the other three come out right (generally speaking, of course).

This still leaves a problem for the actual OpenML calculated value though, with a mean kurtossis of 32.79.

**Comment by berndbischl** [comment]
> This still leaves a problem for the actual OpenML calculated value though, with a mean kurtossis of 32.79.

yes and no, 
it also leaves the 2nd problem why the 2 data sets are not the same! i dont see any reason at all how they could or should be different. i mean it's "iris"....

**Comment by berndbischl** [comment]
ah cool

from wikipedia

https://en.wikipedia.org/wiki/Iris_flower_data_set

(scroll down)

"Fisher's Iris Data". (Contains two errors which are documented). UCI Machine Learning Repository: Iris Data Set.

(click link)

This data differs from the data presented in Fishers article (identified by Steve Chadwick, spchadwick '@' espeedaz.net ). The 35th sample should be: 4.9,3.1,1.5,0.2,"Iris-setosa" where the error is in the fourth feature. The 38th sample: 4.9,3.6,1.4,0.1,"Iris-setosa" where the errors are in the second and third features.

**Comment by berndbischl** [comment]
OML imported the (erroneous) version from UCI....

**Comment by PG-TUe** [comment]
You beat me to it! I guess it might be good to maybe upload a different version then, as it is so often used for examples.

**Comment by berndbischl** [comment]
@joaquinvanschoren 

great chance to now try out whether its possible to use the "problem here, flag a data set as faulty" mechanism....

i would have just never guessed that we need it already for iris .... :(

**Comment by PG-TUe** [comment]
I suppose the UCI link was in the description of the dataset.. but yeah, I would not have expected that.
Also, I did submit an issue on the site, then removed it (thought I did something wrong), and now I can't submit an issue anymore (well, technically it looks like it lets me submit an issue, but the site doesn't respond).

**Comment by joaquinvanschoren** [comment]
Interesting! Will upload the new version. From which package did you get
the 'correct' iris?

I also noticed a bug in the issue reporting system, will look at this asap
(although I'm on holiday :)).

Cheers,
Joaquin

**Comment by PG-TUe** [comment]
Good to hear, but don't overwork yourself! Anyhow, this still leaves open the issue with calculated kurtosis.

**Comment by PG-TUe** [comment]
Actually, it looks like some other meta-features are not up to snuff either. When we look at dataset 62, the zoo dataset ([link](http://www.openml.org/d/62)), we see it only has one numerical features, the number of legs (probably should have been nominal, but nevermind that fact). We see that this feature only has values 0, 2, 4, 5, 6 and 8. Yet the MeanMeansOfNumericAtts is 100.14.

I also started working on information-theoretic meta-features, and while the ClassEntropy seems correct, my MeanAttributeEntropy feature also differs, though I haven't dived in to check whether or not that is on my end. All of the features that go wrong so far are aggregate statistics, I am not sure if that helps any.

**Comment by berndbischl** [comment]
> Interesting! Will upload the new version. From which package did you get
> the 'correct' iris?

http://www.rdocumentation.org/packages/datasets/versions/3.3.1/topics/iris


> Will upload the new version.

i would be interested in the process of "flagging" the other data set now, what that implies and happens. can you please outline this here? this seems a good usecase now.

**Comment by PGijsbers** [comment]
The mean of means for Zoo is now ~2.3 and kurtosis is -0.75 which more closely matches the initial computation I did. So I am still not sure what went wrong, but it seems that the meta feature computation might now work correctly.

@SubhadityaMukherjee be aware that at least at some point in time, there were some incorrect computations.

Issue of the flagging feature remains unaddressed, but wasn't the point of this issue. Having a way to discuss/flag/correct datasets is something that is a known issue (the current practice largely relies on the openml/openml-data repo).
